### PR TITLE
Adding more map scope crs items

### PR DESCRIPTION
**Pull request by roya0045**
## Description
Often when making maps I find that there is a lack of CRS variables to properly explain the CRS used in the map.

So I decided to add some new variables to the QgsLayoutItemMap ExpressionScope.

This modifies one definition (since there is a duplicate), the name change of map_definition to map_crs_Proj4 could be duplicated to prevent confusion but I think mentioning that the output is Proj4 makes things more clear.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Review by nyalldawson — COMMENTED** [review]
Looks handy!

You'll need to:

1. Copy the new variables to https://github.com/qgis/QGIS/blob/master/src/core/qgsexpressioncontext.cpp#L954 -- unfortunately theres some duplicate code here, for reasons described at the top of QgsExpressionContextUtils::mapSettingsScope

2. We need new unit tests covering the new variables in https://github.com/qgis/QGIS/blob/13e71864939b4a24b6a73811afec56bf34f42a97/tests/src/core/testqgsmapsettings.cpp#L31 and https://github.com/qgis/QGIS/blob/master/tests/src/core/testqgslayoutmap.cpp#L591

File: src/core/expression/qgsexpression.cpp [hunk 1]
```diff
@@ -745,8 +745,12 @@ void QgsExpression::initVariableHelp()
   sVariableHelpTexts.insert( QStringLiteral( "map_extent_width" ), QCoreApplication::translate( "variable_help", "Width of map." ) );
   sVariableHelpTexts.insert( QStringLiteral( "map_extent_height" ), QCoreApplication::translate( "variable_help", "Height of map." ) );
   sVariableHelpTexts.insert( QStringLiteral( "map_crs" ), QCoreApplication::translate( "variable_help", "Coordinate reference system of map (e.g., 'EPSG:4326')." ) );
-  sVariableHelpTexts.insert( QStringLiteral( "map_crs_definition" ), QCoreApplication::translate( "variable_help", "Coordinate reference system of map (full definition)." ) );
+  sVariableHelpTexts.insert( QStringLiteral( "map_crs_description" ), QCoreApplication::translate( "variable_help", "Name of theCoordinate reference system of the map." ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpression.cpp]:
```suggestion
  sVariableHelpTexts.insert( QStringLiteral( "map_crs_description" ), QCoreApplication::translate( "variable_help", "Name of the coordinate reference system of the map." ) );
```

File: src/core/expression/qgsexpression.cpp [hunk 2]
```diff
@@ -745,8 +745,12 @@ void QgsExpression::initVariableHelp()
...
   sVariableHelpTexts.insert( QStringLiteral( "map_units" ), QCoreApplication::translate( "variable_help", "Units for map measurements." ) );
+  sVariableHelpTexts.insert( QStringLiteral( "map_crs_acronym" ), QCoreApplication::translate( "variable_help", "Acronym of the Coordinate reference system of the map." ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpression.cpp]:
```suggestion
  sVariableHelpTexts.insert( QStringLiteral( "map_crs_acronym" ), QCoreApplication::translate( "variable_help", "Acronym of the coordinate reference system of the map." ) );
```

File: src/core/expression/qgsexpression.cpp [hunk 3]
```diff
@@ -745,8 +745,12 @@ void QgsExpression::initVariableHelp()
...
+  sVariableHelpTexts.insert( QStringLiteral( "map_ellipsoid_acr" ), QCoreApplication::translate( "variable_help", "Acronym of the ellipsoid of theCoordinate reference system of the map." ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpression.cpp]:
```suggestion
  sVariableHelpTexts.insert( QStringLiteral( "map_ellipsoid_acr" ), QCoreApplication::translate( "variable_help", "Acronym of the ellipsoid of the coordinate reference system of the map." ) );
```

File: src/core/expression/qgsexpression.cpp [hunk 4]
```diff
@@ -745,8 +745,12 @@ void QgsExpression::initVariableHelp()
...
+  sVariableHelpTexts.insert( QStringLiteral( "map_crs_Proj4" ), QCoreApplication::translate( "variable_help", "Proj4 description of the Coordinate reference system." ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpression.cpp]:
```suggestion
  sVariableHelpTexts.insert( QStringLiteral( "map_crs_Proj4" ), QCoreApplication::translate( "variable_help", "Proj4 description of the coordinate reference system." ) );
```

File: src/core/expression/qgsexpression.cpp [hunk 5]
```diff
@@ -745,8 +745,12 @@ void QgsExpression::initVariableHelp()
...
+  sVariableHelpTexts.insert( QStringLiteral( "map_crs_Wkt" ), QCoreApplication::translate( "variable_help", "Wkt description of the Coordinate reference system." ) );
```
Comment (nyalldawson) [inline src/core/expression/qgsexpression.cpp]:
```suggestion
  sVariableHelpTexts.insert( QStringLiteral( "map_crs_Wkt" ), QCoreApplication::translate( "variable_help", "WKT definition of the coordinate reference system." ) );
```

**Comment by roya0045** [comment]
> Looks handy!
> 
> You'll need to:
> 
> 1. Copy the new variables to https://github.com/qgis/QGIS/blob/master/src/core/qgsexpressioncontext.cpp#L954 -- unfortunately theres some duplicate code here, for reasons described at the top of QgsExpressionContextUtils::mapSettingsScope
> 2. We need new unit tests covering the new variables in 
>      
>        
>          [QGIS/tests/src/core/testqgsmapsettings.cpp](https://github.com/qgis/QGIS/blob/13e71864939b4a24b6a73811afec56bf34f42a97/tests/src/core/testqgsmapsettings.cpp#L31)
>        
>        
>             Line 31
>          in
>          [13e7186](/qgis/QGIS/commit/13e71864939b4a24b6a73811afec56bf34f42a97)
>        
>        
>        
>        
>    
>            
>              
>               class TestQgsMapSettings: public QObject 
>            
>        
>      
>    
>     and https://github.com/qgis/QGIS/blob/master/tests/src/core/testqgslayoutmap.cpp#L591

were these comments made considering that I already added elements to the QgsExpression and made some test case or do I need to add more?

**Comment by nyalldawson** [comment]
There's an unfortunate duplication of code - you need to also add the new variables to the mapSettingsScope from https://github.com/qgis/QGIS/blob/master/src/core/qgsexpressioncontext.cpp#L954, and then add unit tests covering *all* the newly added variables (for both layout maps and the mapSettingsScope)

**Comment by roya0045** [comment]
@nyalldawson I'll try to work on the unit test tomorrow if I have some free time.

**Comment by roya0045** [comment]
Unit tests are done.

**Review by nyalldawson — APPROVED** [review]
Looks good! I'll tag this for a merge when master is unfrozen again

**Comment by roya0045** [comment]
@nyalldawson I am clueless as to why the map_crs_definition tests are failing, any guess?

**Comment by roya0045** [comment]
I just did the toProj4() on my 3.2.3 and it returns an empty string, just like the current tests return an empty string. This raises the question as to how did the tests even pass previously. ( I tried with epsg 4326 and another random one.)

File: tests/src/core/testqgslayoutmap.cpp [hunk 6]
```diff
@@ -626,13 +626,13 @@ void TestQgsLayoutMap::expressionContext()
   r = e6.evaluate( &c );
   QCOMPARE( r.toString(), QString( "longlat" ) );
 
-  QgsExpression e7( QStringLiteral( "@map_crs_pro4" ) );
-  r = e7.evaluate( &c );
-  QCOMPARE( r.toString(), QString( "+proj=longlat +datum=WGS84 +no_defs" ) );
+  //QgsExpression e7( QStringLiteral( "@map_crs_pro4" ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgslayoutmap.cpp]: The issue here is a type - "pro4", not "proj4"!

**Comment by nyalldawson** [comment]
@roya0045 I've pulled your work into #8933, mainly to get the new unit tests so that I could extend them to cover the scale work. FYI : I renamed the ellipsoid acronym variable to avoid any ambiguity, as I plan on adding a @map_ellipsoid variable at some later stage.

File: tests/src/core/testqgslayoutmap.cpp (hunk 6 above)
Comment (roya0045) [reply tests/src/core/testqgslayoutmap.cpp]: I thought it would return an error when an item doesn't exist rather than an empty string. :/

**Pull request closed without merging**